File: Algorithms/173.BinarySearchTreeIterator.py
class BSTIterator:
    # @param root, a binary search tree's root node
    def __init__(self, root):
        self.path = []
        while root:
            self.path.append(root)
            root = root.left

    # @return a boolean, whether we have a next smallest number
    def hasNext(self):
        return len(self.path) != 0
        

    # @return an integer, the next smallest number
    def next(self):
        #print 'path:', self.path[-1].val
        result = self.path[-1]
        self.path.pop()
        temp = result.right
        while temp:
            self.path.append(temp)
            temp = temp.left
        return result.val

File: Algorithms/cpp/173.BinarySearchTreeIterator.cpp
class BSTIterator {
private:
    stack<TreeNode *> path;
public:
    BSTIterator(TreeNode *root) {
        while (root != NULL) {
            path.push(root);
            root = root->left;
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return path.size() != 0;
    }

    /** @return the next smallest number */
    int next() {
        TreeNode *result = path.top();
        path.pop();
        TreeNode *temp = result->right;
        while (temp != NULL) {
            path.push(temp);
            temp = temp->left;
        }
        return result->val; 
    }
};
